pkg/kubelet/cm/deviceplugin/endpoint.go

```diff
@@ -95,91 +102,60 @@ func (e *endpointImpl) getDevices() []pluginapi.Device {
 // device states with its cached states to get list of new, updated, and deleted devices.
 // It then issues a callback to pass this information to the device manager which
 // will adjust the resource available information accordingly.
-func (e *endpointImpl) run() {
+func (e *endpointImpl) Run() {
+	glog.V(3).Infof("Starting to run endpoint %s", e.resourceName)
+	e.stopChan = make(chan interface{}, 0)
+
 	stream, err := e.client.ListAndWatch(context.Background(), &pluginapi.Empty{})
 	if err != nil {
 		glog.Errorf(errListAndWatch, e.resourceName, err)
-
 		return
 	}
 
-	devices := make(map[string]pluginapi.Device)
-
-	e.mutex.Lock()
-	for _, d := range e.devices {
-		devices[d.ID] = d
-	}
-	e.mutex.Unlock()
-
 	for {
 		response, err := stream.Recv()
 		if err != nil {
-			glog.Errorf(errListAndWatch, e.resourceName, err)
-			return
-		}
-
-		devs := response.Devices
-		glog.V(2).Infof("State pushed for device plugin %s", e.resourceName)
+			s := e.Store()
 
-		newDevs := make(map[string]*pluginapi.Device)
-		var added, updated []pluginapi.Device
+			s.Callback(e.resourceName, nil, nil, s.Devices())
+			e.clientConn.Close()
 
-		for _, d := range devs {
-			dOld, ok := devices[d.ID]
-			newDevs[d.ID] = d
-
-			if !ok {
-				glog.V(2).Infof("New device for Endpoint %s: %v", e.resourceName, d)
-
-				devices[d.ID] = *d
-				added = append(added, *d)
-
-				continue
-			}
-
-			if d.Health == dOld.Health {
-				continue
-			}
-
-			if d.Health == pluginapi.Unhealthy {
-				glog.Errorf("Device %s is now Unhealthy", d.ID)
-			} else if d.Health == pluginapi.Healthy {
-				glog.V(2).Infof("Device %s is now Healthy", d.ID)
-			}
-
-			devices[d.ID] = *d
-			updated = append(updated, *d)
-		}
-
-		var deleted []pluginapi.Device
-		for id, d := range devices {
-			if _, ok := newDevs[id]; ok {
-				continue
-			}
-
-			glog.Errorf("Device %s was deleted", d.ID)
+			glog.Errorf(errListAndWatch, e.resourceName, err)
 
-			deleted = append(deleted, d)
-			delete(devices, id)
+			close(e.stopChan)
+			return
 		}
 
-		e.mutex.Lock()
-		e.devices = devices
-		e.mutex.Unlock()
+		glog.V(2).Infof("Endpoint %s updated", e.resourceName)
 
-		e.callback(e.resourceName, added, updated, deleted)
+		s := e.Store()
+		added, updated, deleted := s.Update(response.Devices)
+		s.Callback(e.resourceName, added, updated, deleted)
 	}
 }
 
 // allocate issues Allocate gRPC call to the device plugin.
-func (e *endpointImpl) allocate(devs []string) (*pluginapi.AllocateResponse, error) {
+func (e *endpointImpl) Allocate(devs []string) (*pluginapi.AllocateResponse, error) {
 	return e.client.Allocate(context.Background(), &pluginapi.AllocateRequest{
 		DevicesIDs: devs,
 	})
 }
 
-func (e *endpointImpl) stop() {
+func (e *endpointImpl) Stop() error {
 	e.clientConn.Close()
+
+	if e.stopChan == nil {
+		return nil
+	}
+
+	select {
+	case <-e.stopChan:
```

Comment:
in what case, this `case` will be selected?

Reply:
See previous comment